nroff/nroffdoc/nrinfo.1
.Id $Id: nrinfo.1,v 1.1 2000-10-21 14:31:35 jon Exp $
.Im mantools utility
.TH nrinfo 1
.XI nrinfo - "Nroff/Troff subject category extraction utility"
.SH NAME
nrinfo \- Nroff/Troff subject category extraction utility
.SH SYNOPSIS
.sp 1
.na
\fBnrinfo\fR
\fC[\fB\-?\fC|\fB\-h\fC]\fR
\fC[\fB\-E\fR\ \fI<file>\fC]\fR
\fC[\fB\-e\fR\ \fI<file>\fC]\fR
\fC[\fB\-I\fC]\fR
\fC[\fB\-o\fR\ \fI<file>\fC]\fR
\fC[\fB\-q\fC]\fR
\fInroff\ files...\fR
.ad
.SH DESCRIPTION
\fBnrinfo\fR  is a Nroff/Troff  utility used to determine all of the component
tags in the nroff files supplied on the command line. The utility  generates a
list of all  component  names that have been assigned and is used to check for
consistency.  A single line of output is produced for each  component tag that
has been  encountered,  where  multiple tags of the same name are defined then
only the first entry that is  encountered  is reported. A typical  output line
would appear as follows:
.PP
.na
.RS
Module [\fI<moduleName>\fR].
Component [\fI<componentName>\fR]:
\fI<fileName>\fR
.PP
.RE
.ad
The fields are defined as follows:
.IP "\fI<moduleName>\fR"
The  name of the  module  this is  determined  from  the  first  field  of the
\fB\.Im\fR field defined by
.Ht mad 7 .
.IP "\fI<componentName>\fR"
The name of the  component.  This is  determined  by the  second  field of the
\fB\.Im\fR field in addition to the \fB\.XI\fR files, as defined by
.Ht mad 7 .
.IP "\fI<fileName>\fR"
The name of the first file  encountered  that  contained  a  reference  to the
specified tag.
.PP
The utility also detects tags that have been assigned  incorrectly and reports
an error if any NULL tags are encountered. A NULL tag is only reported when an
\fIor\fR  option  (denoted  by a  \fC|\fR)  has been  assigned  and one of the
entries is  undefined.  Component  names  which are omitted in the  \fB\.Im\fR
field  default  to  the  \fImisc\fR  component.   Components  omitted  in  the
\fB\.XI\fR field are optional and are not assigned a default component.
.PP
The utility by default  generates  output on  \fIstdout\fR.  The output may be
diverted to a \fIfile\fR  using the \fB-o\fR  option on the command  line. The
command line options are defined as follows:-
.\" -?/-h help option
.so opt_h.tni
.\" -e -E : Error options.
.so opt_e.tni
.\" -I : Information.
.so opt_i.tni
.\" -o <file> : Output file name.
.so opt_o.tni
.\" -q : Quiet mode.
.so opt_q.tni
.SH EXAMPLE
The  following  output  (partial) was  generated  using the following  command
line:-
.CS
nrinfo -o vc.nri vcm3n*.3
.CE
The output has been reduced for the purposes of this manual page.
.CS
.ne 6
\&ERROR: vcm3n108.3: 14: Null tag
\&ERROR: vcm3n108.3: 15: Null tag
Module [vc]. Component [actor;init]: vcm3n006.3
Module [vc]. Component [attribute;callback]: vcm3n012.3
Module [vc]. Component [attribute;construct]: vcm3n020.3
Module [vc]. Component [attribute;misc]: vcm3n049.3
Module [vc]. Component [attribute;traverse]: vcm3n042.3  <<< ERROR
Module [vc]. Component [attribute;userdata]: vcm3n049.3
Module [vc]. Component [attributeitem]: vcm3n002.3
Module [vc]. Component [attributel;traverse]: vcm3n048.3 <<< ERROR
Module [vc]. Component [attributritem]: vcm3n062.3
Module [vc]. Component [audio;attribute]: vcm3n002.3
Module [vc]. Component [audio;radiator]: vcm3n089.3
Module [vc]. Component [audio;resource;applic]: vcm3n038.3
Module [vc]. Component [audio;resource;server]: vcm3n039.3
Module [vc]. Component [body;attribute]: vcm3n040.3
Module [vc]. Component [body;collide]: vcm3n102.3
Module [vc]. Component [body;constraints]: vcm3n034.3
Module [vc]. Component [body;flying]: vcm3n032.3
:::
Module [vc]. Component [visual;resource;server]: vcm3n135.3
Module [vc]. Component [visual;resource;view]: vcm3n094.3
Module [vc]. Component [xwindowid]: vcm3n088.3
Module [vc]. Component [zone]: vcm3n062.3
.CE
The \fCERROR\fR  lines indicate that a tag has been incorrectly  assigned, and
is generally an indication  that an \fIor\fR option (denoted by a \fC|\fR) has
been assigned and one of the entries is undefined.
.PP
The  output is used may be used to detect  inconsistent  or  misspelled  tags.
Consider the previous example at lines 11 and 14 marked with
.C "<<<\ ERROR"
The tags are  quite  obviously  are  supposed  to be the  same but  line 14 is
spelled  incorrectly.  The first  occurrence of this  mis-spelling is the file
\fIvcm3n048.3\fR.
.SH "SEE ALSO"
.na
.Hr groff 1 ,
.Hr nawk 1 ,
.Hr nroff 1 ,
.Ht nr2html 1 ,
.Ht nr2rtf 1 ,
.Ht nrcheck 1 ,
.Ht nrsearch 1 ,
.Ht sm 4 .
.ad
.FH


nroff/nroff/nrinfo.c
/****************************************************************************
 *
 *			Copyright 1997 Jon Green.
 *			    All Rights Reserved
 *
 *
 *  System        :
 *  Module        :
 *  Object Name   : $RCSfile: nrinfo.c,v $
 *  Revision      : $Revision: 1.2 $
 *  Date          : $Date: 2000-10-21 15:02:02 $
 *  Author        : $Author: jon $
 *  Last Modified : <001021.1405>
 *
 *  Description
 *
 *  Notes
 *
 *  History
 *
 *  $Log: not supported by cvs2svn $
 *  Revision 1.1  2000/10/21 14:31:29  jon
 *  Import
 *
 *
 ****************************************************************************
 *
 *  Copyright (c) 1997 Jon Green.
 *
 *  All Rights Reserved.
 *
 *  This Document may not, in whole or in part, be copied,
 *  photocopied, reproduced, translated, or reduced to any
 *  electronic medium or machine readable form without prior
 *  written consent from Jon Green.
 ****************************************************************************/

static const char rcsid[] = "@(#) : $Id: nrinfo.c,v 1.2 2000-10-21 15:02:02 jon Exp $";

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <ctype.h>
#include <signal.h>

#if ((defined _HPUX) || (defined _LINUX))
#include <unistd.h>
#else
#include <getopt.h>
#endif

#include <utils.h>
#include "nroff.h"

/* Macro Definitions */
/*
 * 1.0.0a - JG 16/11/96 Integrated new utilies library.
 * 1.0.0b - JG 05/12/95 Added bullet support.
 * 1.0.0c - JG 27/09/97 Added package support.
 * 1.0.0d - JG 03/05/97 Ported to win32.
 */

#define MODULE_VERSION  "1.0.0d"
#define MODULE_NAME     "nrinfo"

static char *sectionId = NULL;
static char *sectionModule = NULL;
static char *sectionComponent = NULL;
static char *progname = MODULE_NAME;
static FILE *fpr = NULL;

/*
 * addComponent
 * Add a component to the tags. Decompose into it's component parts.
 */


static void
addComponent (char *s)
{
    char *allocBuf;                     /* Allocated buffer */
    char *q;                            /* Start of current component */
    char *p;                            /* Temporary character string */
    char c;                             /* Current charater */

    allocBuf = bufNStr (NULL, s);       /* Copy string */
    if ((q = allocBuf) == NULL)         /* Quit if NULL */
        return;

    p = q;
    for (;;)
    {
        if (((c = *p) == '\0') || (c == '|'))
        {
            *p = '\0';
            if (*q == '\0')
                uError ("Null tag\n");
            else
            {
                TagP tp;

                /*
                 * Allocate a new tag and add to list.
                 */

                tp = tagAlloc (sectionModule, q, NULL, nrfp->fileName,
                               nrfp->lineNo, NULL);
                add_tag (NULL, tp);
            }
            if (c == 0)
                break;
            else
                q = ++p;                /* Reset start of tag */
        }
        else
            p++;                        /* Next character */
    }

    bufFree (allocBuf);                 /* Free off allocated buffer */
}

static void
nrIm_func (char *module, char *component)
{
    bufFree (sectionModule);
    if ((sectionModule = bufNStr (NULL, module)) == NULL)
        sectionModule = bufNStr (NULL, "UNDEFINED");
    bufFree (sectionComponent);
    sectionComponent = bufNStr (NULL, ((component == NULL) ? "misc" : component));
    addComponent (sectionComponent);
}

static void
nrTH_func (char *id, char *num, char *date, char *company, char *title)
{
    bufFree (sectionId);
    sectionId = bufStr (NULL, num);
}

static void
nrStartInc (char *s, int *imode)
{
    *imode = 1;
}

static void
nrXI_func (char *name, char *id, char *desc, char *comp)
{
    addComponent (comp);
}

static void
nrXJ_func (char *name, char *id, char *desc, char *comp)
{
    addComponent (comp);
}

static void
alpha_term (void)
{
    TagP lp;
    TagP tp;

    lp = NULL;
    for (tp = tagIterateInit (NULL); tp != NULL; tp = tagIterate (tp)) {
        if (lp != NULL) {
            if ((strcmp (lp->name, tp->name) == 0) &&
                (strcmp (lp->section, tp->section) == 0))
                lp = tp;
        }
        if (lp != tp)
            fprintf ((fpr == NULL) ? stdout : fpr, "Module [%s]. Component [%s]: ", &tp->name[1], tp->section);
        fprintf ((fpr == NULL) ? stdout : fpr, "%s\n", tp->file);
        lp = tp;
    }
}

static void
usage (void)
{
    fprintf (stdout, "%s : Usage %s [Options] <files>\n", progname, progname);
    fprintf (stdout, "\n");
    fprintf (stdout, "-?/h      : Help - this page\n");
    fprintf (stdout, "-E <file> : Log errors to <file> (append)\n");
    fprintf (stdout, "-e <file> : Log errors to <file> (re-write)\n");
    fprintf (stdout, "-I        : Version Information\n");
    fprintf (stdout, "-o <file> : Output to <file> default is <files>.out\n");
    fprintf (stdout, "-q        : Quiet mode\n");
    exit (1);
}

static void
infoInitialise (void)
{
    static  nrFUNCTION funcTab = {NULL};

    /* Headers */
    nrInstall (funcTab, TH_func, nrTH_func);

    /* Identification */
    nrInstall (funcTab, XI_func, nrXI_func);
    nrInstall (funcTab, XJ_func, nrXJ_func);
    nrInstall (funcTab, Im_func, nrIm_func);

    /* Files */
    nrInstall (funcTab, startFile_func, NULL);
    nrInstall (funcTab, includeFile_func, nrStartInc);
    nrInstall (funcTab, endFile_func, NULL);

    nrInstallFunctionTable (&funcTab);
}

int
main (int argc, char *argv [])
{
    char    *oname = NULL;              /* Output file name */
    int     ecount = 0;                 /* Error count */
    int     wcount = 0;                 /* Warn count */
    nrFILE  *nrfp;
    int     c;
    int     i;

    /* Initialise the error channel */
    infoInitialise ();
    uErrorSet (0, &ecount);             /* Max Num entries + counter */
    uWarnSet (&wcount);                 /* Warning count */
    uUtilitySet (progname);             /* Set name of program */

    while (1) {
        c = getopt (argc, argv, "e:E:?hqIo:");
        if (c == EOF)
            break;
        switch (c) {
        case 'e':
            uErrorChannelSet (optarg, UERROR_LOG_CREATE);
            break;
        case 'E':
            uErrorChannelSet (optarg, UERROR_LOG_APPEND);
            break;
        case 'I':
            fprintf (stdout, "%s. Version %s. %s. %s\n",
                     progname, MODULE_VERSION, __DATE__, nroffVersion);
            exit (0);
            break;
        case 'q':
            uInteractiveSet (0);        /* Disable interactive mode */
            break;
        case 'o':
            oname = optarg;
            break;
        case '?':
        case 'h':
            usage ();
            break;
        default:
            uError ("Cannot understand option [-%c]\n", c);
            usage();
            break;
        }
    }

    uOpenErrorChannel ();
    argv = getfiles (&argc, argv, optind);

    if (argc > 1) {
        if (oname != NULL) {
            if ((fpr = fopen (oname, "wb")) == NULL)
                uFatal ("Cannot open file [%s]\n", oname);
            uVerbose (0, "Output in file [%s].\n", oname);
        }
        else
            fpr = NULL;
    }
    else
        fpr = NULL;

    /* Process all of the files */

    for (i = 1; i < argc; i++) {
        if ((nrfp = nrFilePush (argv[i])) == NULL)
            exit (1);

        /* Set up the conversion */
        nroff (nrfp, NROFF_MODE_DEFAULT);
    }
    alpha_term ();
    if (fpr != NULL)
        fclose (fpr);

    uCloseErrorChannel ();
    return (ecount);
}
